### PR TITLE
Lkt: check for @invalid annotations in lkt_toolbox

### PR DESCRIPTION
**Pull request by thvnx**
lkt_toolbox now raise an error if a declaration annotated with
@invalid doesn't emit any diagnostic, and conversely, it also checks
that a diagnostic is related to an @invalid declaration.

TN: UB03-005

### CODE REVIEW

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:65]: This lacks a description of what the maps map to (i.e. the meaning of boolean there)

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:61]: I would suggest to directly mention the map variable that this populates, and be more specific “invalid declaration” is a bit confusing: a decl with the @invalid annotation versus a declaration that is actually invalid: we can’t assume there that they are the same, otherwise lkt_toolbox would work only on correct tests, while the point here is to check test correctness.
```
Populate ``Invalid_Decl_Map`` and reject declarations with ``@invalid``
annotations that are nested in another declaration annotated with ``@invalid``.
```

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:99]: Please refactor to avoid the `Analysis.Node (S)` repetitions

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:129]: We generally put `is` on its own line when there are declarations before the `begin` keyword (less confusing layout)

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:60]: There is one extra indentation space: as discussed live earlier, we add only 2 columns when the next line is an opening paren. (same multiple times below)

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:144]: The other map operation above does not need the `As_Lk_Node` conversion: why is it needed here? (same below)

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:194]: Please move `then` on the line above

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:195]: Same remark as above wrt. the exception

File: contrib/lkt/language/parser.py
Comment (pmderodat) [inline contrib/lkt/language/parser.py:479]: Can you document what the `topmost` argument is? Also clarify whether the current node is included in the search (it looks like it’s not looking at the implementation, but that should be documented here).

**Review by pmderodat — CHANGES_REQUESTED** [review]

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (thvnx) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: Should be good now but I couldn't add some tests for those new errors since lkt_toolbox returns a non-zero status code.

Should I add some? This will require to add a new lkt_toolbox option to set the status code, and to modify lkt_driver to manage arguments.

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: As discussed live the other day, you can modify `Node`'s declaration to be a simple `LK_Node`, and initialize it with `Analysis.Node (S).As_LK_Node`.

**Review by pmderodat — CHANGES_REQUESTED** [review]

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: I think it’s fine not to add tests, but it’s for sure best to have some: your call!

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (thvnx) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: I still end up with a compilation error when trying to get rid of this `As_LK_Node` conversion [here](https://github.com/AdaCore/langkit/blob/f87bb05f218fc176d8986d872e72c2238886f164/contrib/lkt/extensions/mains/lkt_toolbox.adb#L177). As it is used in the traversal function, `Node` should be of type `LK_Node'Class` no? I guess we should discuss it again.

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (thvnx) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: I've added a new procedure to take care of that. Thanks for the suggestion.

File: contrib/lkt/language/parser.py
Comment (thvnx) [reply contrib/lkt/language/parser.py:482]: Actually, it does start from `Entity` (and therefore can return it): it starts by looking at the parent in order to start the recursion from it, by setting `topmost` to `Entity` if it is an `@invalid` declaration. There is only one case where it cannot return `Entity`: when it has no parent, i.e. `Entity` is the root node, which is OK since it cannot be a declaration annotated with `@invalid`.
```
        return Entity.parent.then(
            lambda p: p.topmost_invalid_decl(
                Entity.is_a(FullDecl).then(
                    lambda _: Entity.cast(FullDecl)
                    .has_annotation('invalid').then(
                        lambda _: Entity,             # <- starts from Entity
                        default_val=topmost
                    ),
                    default_val=topmost
                )
            ),
            default_val=topmost
        )
```

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: That’s correct: Node being a `LK_Node'Class` value because of the `Traverse` subprogram (and you can’t change that), you need to call `.As_LK_Node` to pass it as a `LK_Node` value.

It seems that GitHub won’t let me see the original commit for which I wrote “same below”, so I can’t double check: sorry if I was misleading. The simplification was valid for `Node.P_Topmost_Invalid_Decl` in any case.

File: contrib/lkt/language/parser.py
Comment (pmderodat) [reply contrib/lkt/language/parser.py:482]: Oh I see, you are correct, thanks. I find this specific recursion hard to proof-read, there are odd repetitions (`default_val=topmost` appears 3 times) and the fact that the topmost argument (which is there only as an implementation detail) leaks to the user API is a smell. I would suggest to use `.parents()` in order to switch to an iterative (and straightforward) form. Something such as:

```python
return Self.parents().filter(
    lambda p:
    p.cast(FullDecl).then(lambda fd: fd.has_annotation("invalid"))
).at(-1)
```

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (thvnx) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: Will keep that for later then!

File: contrib/lkt/language/parser.py
Comment (thvnx) [reply contrib/lkt/language/parser.py]: Done!

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:169]: declarations*

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:62]: It looks like all calls to this procedure use `To_Unbounded_Text`: what about making it part of the factorization too?

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:125]: This will probably get refactored, but: this still has the source layout issue I mentionned in https://github.com/AdaCore/langkit/pull/578#discussion_r748436038

**Review by pmderodat — CHANGES_REQUESTED** [review]
LGTM after minor nits. Thanks!

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (thvnx) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb]: Done!

File: contrib/lkt/extensions/mains/lkt_toolbox.adb
Comment (pmderodat) [inline contrib/lkt/extensions/mains/lkt_toolbox.adb:126]: Not quite: should be

```
Print_Lkt_Toolbox_Diagnostic
  (Node, "unexpected diagnostic, is @invalid annotation missing?");
```
In general: paren subexpressions should not go backwards in column numbers compared to the opening paren.

Comment (thvnx) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb:127]: Thanks! Hope it's good now!

Comment (pmderodat) [reply contrib/lkt/extensions/mains/lkt_toolbox.adb:127]: Yes, thanks!

**Review by pmderodat — APPROVED** [review]

**Pull request merged**
